### PR TITLE
Support for modal time integration

### PR DESCRIPTION
**Pull request by kmokstad**
The first two commits are more of cosmetic character.
Then a new system matrix class for diagonal systems is added.
which is used to facilitate the assembly of modal systems.

### CODE REVIEW

**Comment by kmokstad** [comment]
The WIP tag applies to the last commit only

**Comment by kmokstad** [comment]
jenkins build this with downstreams please

**Comment by kmokstad** [comment]
jenkins build this with downstreams please

**Comment by kmokstad** [comment]
jenkins build this with downstreams please

**Comment by kmokstad** [comment]
jenkins build this with downstreams please

**Comment by kmokstad** [comment]
I rest my case, don't know why this failure generates @akva2

**Comment by kmokstad** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-stokes=124 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-stokes=125 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=94 ifem-finitedeformation=24 ifem-beamex=22 ifem-stokes=125 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=94 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=94 please

File: src/LinAlg/DiagMatrix.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+//! \brief Diagonal system matrix representation.
+//!
+//==============================================================================
+
+#ifndef _DIAG_MATRIX_H
+#define _DIAG_MATRIX_H
+
+#include "SystemMatrix.h"
+
+class SparseMatrix;
```
Comment (akva2) [inline src/LinAlg/DiagMatrix.h]: necessary?

File: src/LinAlg/SparseMatrix.C [hunk 2]
```diff
@@ -1385,11 +1390,12 @@ void SparseMatrix::calcCSR(IntVec& IA, IntVec& JA,
   IA.resize(nrow+1,nnz);
   JA.resize(nnz);
 
-  IA[0] = 0;
+  IA.front() = 0;
   size_t cur_row = 1, ix = 0;
-  for (auto it = elem.begin(); it != elem.end(); ++it, ix++) {
-    JA[ix] = it->first.second-1;
-    while (it->first.first > cur_row)
+  for (const ValueMap::value_type& it : elem)
```
Comment (akva2) [inline src/LinAlg/SparseMatrix.C]: it should imo be reserved for iterators.

File: src/LinAlg/SystemMatrix.h [hunk 3]
```diff
@@ -279,6 +279,9 @@ class SystemMatrix
   //! \brief Truncates all small matrix elements to zero.
   virtual bool truncate(Real = Real(1.0e-16)) { return false; }
 
+  //! \brief Multiplication with a scalar.
+  virtual void mult(Real) {}
```
Comment (akva2) [inline src/LinAlg/SystemMatrix.h]: pure virtual?

File: src/LinAlg/DiagMatrix.h (hunk 1 above)
Comment (kmokstad) [reply src/LinAlg/DiagMatrix.h]: Not at all, a survivor from DenseMatrix which was used as template here.

File: src/LinAlg/SparseMatrix.C (hunk 2 above)
Comment (kmokstad) [reply src/LinAlg/SparseMatrix.C]: Agree, changed to val instead

File: src/LinAlg/SystemMatrix.h (hunk 3 above)
Comment (kmokstad) [reply src/LinAlg/SystemMatrix.h]: OK, then have to implement for SPRMatrix also.

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=94 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=94 please

**Comment by kmokstad** [comment]
jenkins build this with downstreams ifem-elasticity=94 ifem-poroelasticity=44 please

**Pull request merged**